### ISSUE TITLE
Incorrectly rendered EPSG:4326 raster in Equal Earth, depending on output image size

### ISSUE THREAD

**Issue by dbaston** [mapserver/mapserver#5958]
When rendering a raster stored in EPSG:4326 to EPSG:8857 (Equal Earth), values are plotted in incorrect locations for certain ouput image sizes.

Mapfile:
```
MAP
  PROJECTION 
    "init=epsg:4326"
  END

  EXTENT -180 -90 180 90        

WEB
  METADATA
  "ows_enable_request" "*"
  "wms_srs"            "EPSG:4326 EPSG:8857" 
  END
END

LAYER
  NAME "rast"
  TYPE RASTER
  STATUS OFF 

  PROJECTION "init=epsg:4326" END

  DATA "/tmp/data.tif"

  CLASS
    EXPRESSION ([pixel] >= 0)
    STYLE
      COLORRANGE 0 0 0 0 0 255
      DATARANGE 0 60
      RANGEITEM "pixel"
    END
  END
END

END
```

Raster data: [data.zip](https://github.com/mapserver/mapserver/files/4012953/data.zip)

Expected output:

```
./mapserv -nh "QUERY_STRING=map=/home/dan/dev/mapserver/test.map&service=wms&version=1.3.0&request=GetMap&format=png&width=2162&height=1055&crs=EPSG%3A8857&layers=rast&BBOX=-17243959.06221695,-8392927.59846645,17243959.06221695,8392927.59846645&transparent=TRUE" > ok.png
```
![image](https://user-images.githubusercontent.com/6318931/71625711-41225500-2bb7-11ea-9b59-abb821cf9b1d.png)

Incorrect output (image width increased from 2162 to 2163 pixels):

```
./mapserv -nh "QUERY_STRING=map=/home/dan/dev/mapserver/test.map&service=wms&version=1.3.0&request=GetMap&format=png&width=2163&height=1055&crs=EPSG%3A8857&layers=rast&BBOX=-17243959.06221695,-8392927.59846645,17243959.06221695,8392927.59846645&transparent=TRUE" > bad.png
```
![image](https://user-images.githubusercontent.com/6318931/71625706-3b2c7400-2bb7-11ea-804e-994a3b4a8fd0.png)

```
./mapserv -v
# MapServer version 7.5-dev OUTPUT=PNG OUTPUT=JPEG SUPPORTS=PROJ SUPPORTS=AGG SUPPORTS=FREETYPE SUPPORTS=CAIRO SUPPORTS=ICONV SUPPORTS=FRIBIDI SUPPORTS=WMS_SERVER SUPPORTS=WFS_SERVER SUPPORTS=WCS_SERVER SUPPORTS=FASTCGI SUPPORTS=GEOS SUPPORTS=PBF INPUT=JPEG INPUT=POSTGIS INPUT=OGR INPUT=GDAL INPUT=SHAPEFILE
```

**Comment by dbaston** [comment]
The two cases (height=2162 or height=2163) and up hitting a different path around here:
https://github.com/mapserver/mapserver/blob/master/mapresample.c#L976

In the 2162 case, `dfDeltaX` is computed to be positive, and the resampling takes place by recursing several times into subsets of the scanline. In the 2163 case `dfDeltaX` is computed to be negative and the resampling takes place with a single pass on the scanline.

**Comment by dbaston** [comment]
Here's what seems to be going on.

- `msApproxTransformer` wants to transform a scanline in map pixel space (0.5 to 2162.5) to image space (0-719).
- It first uses PROJ to transform the first, middle, and last points in pixel space (0.5, 1081.5, and 2162.5) to image space and gets (719.73, 360, 0.267)
- It then tries to use linear interpolation from the first and last points to approximate the middle point. The intention is to avoid using PROJ to transform all of the points if linear interpolation is accurate enough.
- In this case, the linear interpolation exactly computes the middle point but is wrong for all other points! If we modify the code so that entire scanline is transformed by PROJ, we get values that look like this: `719.73, 0.065, 0.398, ... 719.26, 219.60, 719.93, 0.267`. So the delta estimated by linear interpolation from the first and last points has the wrong sign, and all intermediate points (except the middle one!) are incorrect. And this only happens when the first and last points in the scanline are extreme enough, which is why we only see it for output image widths >= 2163 pixels.

A workaround is to disable linear interpolation by initializing the approximate transformer with a negative acceptable error on [line 1768](https://github.com/mapserver/mapserver/blob/master/mapresample.c#L1768).

**Comment by dbaston** [comment]
This particular case could be fixed by using PROJ to transform five points instead of three and then checking that, in addition to a successful interpolation of the center point, the sequence of test points is monotonic. But I'm not sure how robust of a fix that is.

**Comment by dbaston** [comment]
I think this was closed by mistake, @rouault ?

**Comment by rouault** [comment]
> I think this was closed by mistake, @rouault ?

yes, sorry. I've used a wrong ticket number in the commit message of a unrelated fix. Re-opening

**Comment by rouault** [comment]
@dbaston Great analysis of this issue by the way. Perhaps a potential solution would be not to use exactly the middle point, but a point at 45% of the segment . While not being fully bullet proof, it could probably be good enough.

**Comment by dbaston** [comment]
@jmckenna can you please re-open? The commit message in 5f89b67 references this issue by accident.

**Comment by jmckenna** [comment]
This problem still exists with master (for the test case EPSG:8857 it requires a more recent PROJ version - I tried with PROJ 7.1.0)
